genson/src/main/java/com/owlike/genson/stream/JsonWriter.java

```diff
@@ -62,18 +62,7 @@
   private final int _bufferSize = _buffer.length;
   private int _len = 0;
 
-  List<MetadataPair> _metadata = new ArrayList<MetadataPair>();
```

Comment:
I've used initially a list for speed. Is there a particular reason for replacing it by a map when we don't need lookups by name?

Reply:
Yes -- that's the "slightly different fix for `MetadataFeatureTest.testClassMetadataShouldBeSerializedOnceWhenUsingUntypedConverter`" issue I mentioned in the PR description ;-)

Basically, the issue was that the metadata list wasn't cleared in all scenarios (forgot the details, but I remember looking at the issue in debugger), which is why we'd end up with a duplicate `@class` annotation in the list and subsequently in the JSON document.

The original fix for that problem was this flag in `ClassMetadataConverter`
```
skipMetadataSerialization = Wrapper.isOfType(delegate, UntypedConverter.class);
```
but that didn't play nice with some of the evolvability tests, so I decided to fix the root problem and make it impossible to add duplicate metadata in a first place. Basically, the `Map` is used to ensure key uniqueness, not for lookups. I honestly don't think it will have any meaningful impact on performance either way -- it's just a one tiny `Map`, typically containing a single or no entries.

Reply:
I did read your initial PR description but didn't read at the same time the code, once I started reading the code I forgot about the description :)

The root problem here is that we will have two `ClassMetadataConverters` for untyped objects. The first one is the one resolved for`Object.class`, the next one would kick in once the `UntypedConverter` tries to serialize or deserialize (as it will go again through `genson.provideConverter` which will in turn resolve a full chain of converters for the given type, including a `ClassMetadataConverter`).

I think using the map here kind of hides the actual issue. In [this PR](https://github.com/owlike/genson/pull/133) I'm trying to address the different issues we mentioned.

Reply:
Sure, addressing the root issue and eliminating the need to do the work twice is certainly a better option than hiding it :-)

I still think using `List` vs `Map` won't help much performance-wise, considering what it's used for, but am happy to revert the change if that's what you prefer, once #133 is in.